Support configuration using MAC instead of device name (#2622)
bottlerocket-os/bottlerocket: PROVISIONING-METAL.md

#### `net.toml` structure

The configuration file must be valid TOML and have the filename `net.toml`.

> The first and required top level key in the file is `version`; the latest is version `3`.

Author: This will all drastically change when I rebase on the bonding/vlan changes.

Just looking to get eyes on wording, etc.

Reviewer: We probably need to remove line 71 since we now support much more than DHCP4 and DHCP6.

  * `via` (IP address): Gateway IP address.  If no gateway is provided, a scope of `link` is assumed.
  * `route-metric` (integer): Relative route priority.

> Version `3` adds the ability to use a MAC address (colon or dash separated) as the identifier for an interface; this support is limited to interface configuration *only* and may not be used in conjunction with bonds or vlans.
> It also adds support for bonding and vlan tagging.

Reviewer: nit: You refer to bonds and vlans for the first time in a negative light. Ideally we would introduce bonds and vlans and MAC address features, then tell people about them about any specific problems they might have with these features. I don't know if its better to flip it or not, but possibly we would do: 
```suggestion
Version `3` adds support for bonding, vlan tagging, and the ability to use a MAC address (colon or dash separated) as the identifier for an interface.
MAC address identification is limited to interface configuration *only* and may not be used in conjunction with bonds or vlans.
```
I'm fine being overruled too.

* Vlan configuration (map):
  * `kind = "vlan"`: This setting is required to specify a vlan device.
  * `device` (string for device name, not MAC address): Defines the device the vlan should be configured on. If VLAN tagging is required, this device should recieve all IP address configuration instead of the underlying device.
  * `id` (integer): Number between 0 and 4096 specifying the vlan tag on the device

Example `net.toml` version `3` with comments:
=======

Example `net.toml` with comments:

> >>>>>>> 3b2de953 ([PROVISIONING-METAL] Add documentation for using MAC addresses)

Reviewer 2: oops

from = "192.168.14.5"
via = "192.168.14.25"

# Interfaces may be configured using their MAC address rather than the interface name.
# The MAC address must be quoted.
["0e:b3:69:44:b6:33"]
dhcp4 = true

> ["3e:03:69:49:e6:31".static4]

Reviewer 2: So you can't refer to this interface in a bonding or vlan configuration because it has no name?

Reviewer: As far as we could tell, wicked simply won't support it. But we might have missed something, a deep dive into wicked's code could confirm if this is true though.

Author: Confirmed: wicked doesn't support this.
